pytorch_lightning/plugins/training_type/fully_sharded.py

```diff
@@ -0,0 +1,206 @@
+            mixed_precision=precision == "mixed",
+            reshard_after_forward=self.reshard_after_forward,
+            fp32_reduce_scatter=self.fp32_reduce_scatter,
+            compute_dtype=self.compute_dtype,
+            bucket_cap_mb=self.bucket_cap_mb,
+            state_dict_device=self.state_dict_device,
+        ):
+            yield
+
+    def connect(self, model: Module) -> None:
+        super().connect(model)
+        model_call_configure_sharded_model_hook = getattr(model, "call_configure_sharded_model_hook", False)
+        if not model_call_configure_sharded_model_hook:
+            # if model has not called configure sharded model, we reset
+            # the training type plugin's call_configure_sharded_model_hook
+            # to give trainer a chance to configure.
+            self.call_configure_sharded_model_hook = True
+
+    def configure_ddp(self):
+        if not self.cpu_offload:
+            # When using CPU Offload, FSDP will manage the CUDA movement for us.
+            # Note: this would be problematic for large model (which could not fit in one GPU)
+            # as FSDP module.to(device) would first summon all parameters
+            # (TODO: need to figure out solution)
+            self.model_to_device()
+
+        # setup optimizers after fully sharded has wrapped the lightning module
+        self.lightning_module.trainer.accelerator.setup_optimizers(self.lightning_module.trainer)
+
+    def pre_dispatch(self):
+        if self.sync_batchnorm:
+            self.model = self.configure_sync_batchnorm(self.model)
+        self.configure_ddp()
+        self.barrier()
+
+    def model_to_device(self):
+        # ensure we update the device type in the lightning module
+        self.lightning_module.to(self.root_device)
+
+    def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
+        # Currently it is same as default TrainingTypePlugin, i.e. return
+        # the full state dict for FSDP, in the future, we will provide sharded
+        # state dict.
+        return super().lightning_module_state_dict()
+
+    @property
+    def setup_optimizers_in_pre_dispatch(self) -> bool:
+        # Setup optimizers after the Fully Sharded Model has been made
+        return True
+
+    def training_step(self, *args, **kwargs):
+        return self.model.training_step(*args, **kwargs)
+
+    def validation_step(self, *args, **kwargs):
+        return self.model.validation_step(*args, **kwargs)
+
+    def test_step(self, *args, **kwargs):
+        return self.model.test_step(*args, **kwargs)
+
+    def predict_step(self, *args, **kwargs):
+        return self.model.predict_step(*args, **kwargs)
+
+    def post_training_step(self):
+        pass
+
+    @classmethod
+    def register_plugins(cls, plugin_registry: Dict):
+        plugin_registry.register(
+            "fsdp",
+            cls,
+            description="Fully sharded training with checkpointing the full state dict.",
+        )
```

Comment:
for this, by looking at this https://github.com/facebookresearch/fairscale/pull/413

probably we could register one with `reshard_after_forward=False` for speed, maybe call, `fsdp_fast`?

Reply:
After some thought not sure how important this is; currently a user has to wrap their model in nested FSDP `wrap` or `auto_wrap`s, so they probably want `reshard_after_forward` to be set to True; I'm not entirely sure the benefit if you wrap intermediate layers and do not call `reshard_after_forward` to release memory!

Reply:
Could we name it `ddp_fsdp` for consistency or `ddp_fully_sharded` (preferred) ?

Reply:
This already exists in the highest level (in the accelerator connector via the Enum it's called `ddp_fully_sharded`.

But going off https://github.com/PyTorchLightning/pytorch-lightning/pull/7259#discussion_r622364269 I think we should introduce `sdp` `fsdp` as aliases in the plugin registry for sharded DDP and fully sharded DDP, as they kinda match `ddp` better!